src/v/storage/readers_cache.cc

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "storage/readers_cache.h"
+
+#include "model/fundamental.h"
+#include "storage/types.h"
+#include "utils/gate_guard.h"
+#include "utils/mutex.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/lowres_clock.hh>
+
+#include <chrono>
+
+namespace storage {
+
+readers_cache::readers_cache(
+  model::ntp ntp, std::chrono::milliseconds eviction_timeout)
+  : _ntp(std::move(ntp))
+  , _eviction_timeout(eviction_timeout) {
+    monitor_eviction();
+    _probe.setup_metrics(_ntp);
+    // setup eviction timer
+    _eviction_timer.set_callback([this] {
+        _eviction_sem.signal();
+        _eviction_timer.arm(_eviction_timeout);
+    });
+
+    _eviction_timer.arm(_eviction_timeout);
+}
+
+model::record_batch_reader
+readers_cache::put(std::unique_ptr<log_reader> reader) {
+    if (reader->lease_range_base_offset() < model::offset{0}) {
+        // do not cache reader with empty lease
+        return model::record_batch_reader(std::move(reader));
+    }
+    vlog(
+      stlog.trace,
+      "{} - adding reader [{},{}]",
+      _ntp,
+      reader->lease_range_base_offset(),
+      reader->lease_range_end_offset());
+
+    auto ptr = new entry{.reader = std::move(reader)}; // NOLINT
+    _readers.push_back(*ptr);
+    _probe.reader_added();
+    return ptr->make_cached_reader(this);
+}
+
+std::optional<model::record_batch_reader>
+readers_cache::get_reader(const log_reader_config& cfg) {
+    vlog(stlog.trace, "{} - trying to get reader for: {}", _ntp, cfg);
+    auto it = std::find_if(_readers.begin(), _readers.end(), [cfg](entry& e) {
```

Comment:
can you add a comment here about why this is a linear search? 

should the datastructure be a queue? every time you add it you push to the back but often pop from the front ?

Reply:
It is an intrusive list, I will add a comment about linear search